kubernetes/deployments_kubeclient.go

```diff
@@ -0,0 +1,1339 @@
+	apiURL, err := url.Parse(apiURLStr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the hostname (without port) and replace api prefix with prefix arg
+	apiHostname := apiURL.Hostname()
+	if !strings.HasPrefix(apiHostname, "api") {
+		return nil, errs.Errorf("cluster URL does not begin with \"api\": %s", apiHostname)
+	}
+	newHostname := strings.Replace(apiHostname, "api", prefix, 1)
+	// Construct URL using just scheme from API URL, modified hostname and supplied path
+	newURL := &url.URL{
+		Scheme: apiURL.Scheme,
+		Host:   newHostname,
+		Path:   path,
+	}
+	return newURL, nil
+}
+
+func getTimestampEndpoints(metricsSeries ...[]*app.TimedNumberTuple) (minTime, maxTime *float64) {
+	// Metrics arrays are ordered by timestamp, so just check beginning and end
+	for _, series := range metricsSeries {
+		if len(series) > 0 {
+			first := series[0].Time
+			if minTime == nil || *first < *minTime {
+				minTime = first
+			}
+			last := series[len(series)-1].Time
+			if maxTime == nil || *last > *maxTime {
+				maxTime = last
+			}
+		}
+	}
+	return minTime, maxTime
+}
+
+func (kc *kubeClient) getBuildConfigs(space string) ([]string, error) {
+
+	// hook for testing
+	if kc.config.BuildConfigInterface != nil {
+		return kc.config.BuildConfigInterface.GetBuildConfigs(space)
+	}
+
+	// BuildConfigs are OpenShift objects, so access REST API using HTTP directly until
+	// there is a Go client for OpenShift
+
+	// BuildConfigs created by fabric8 have a "space" label indicating the space they belong to
+	queryParam := url.QueryEscape("space=" + space)
+	bcURL := fmt.Sprintf("/oapi/v1/namespaces/%s/buildconfigs?labelSelector=%s", kc.config.UserNamespace, queryParam)
+	result, err := kc.getResource(bcURL, false)
+	if err != nil {
+		return nil, errs.WithStack(err)
+	}
+	// Parse build configs from result
+	kind, ok := result["kind"].(string)
+	if !ok || kind != "BuildConfigList" {
+		return nil, errs.New("no build configs returned from endpoint")
+	}
+	items, ok := result["items"].([]interface{})
+	if !ok {
+		return nil, errs.New("malformed response from endpoint")
+	}
+
+	// Extract the names of the BuildConfigs from the response
+	buildconfigs := []string{}
+	for _, item := range items {
+		bc, ok := item.(map[interface{}]interface{})
+		if !ok {
+			return nil, errs.New("malformed build config")
+		}
+		metadata, ok := bc["metadata"].(map[interface{}]interface{})
+		if !ok {
+			return nil, errs.New("'metadata' object missing from build config")
+		}
+		name, ok := metadata["name"].(string)
+		if !ok || len(name) == 0 {
+			return nil, errs.New("malformed metadata in build config; 'name' is missing or invalid")
+		}
+		buildconfigs = append(buildconfigs, name)
+	}
+	return buildconfigs, nil
+}
+
+func (kc *kubeClient) getEnvironmentsFromConfigMap() (map[string]string, error) {
+	// fabric8 creates a ConfigMap in the user namespace with information on environments
+	const envConfigMap string = "fabric8-environments"
```

Comment:
what about moving those constants out of the function scope ?

Reply:
For future reference, is there a reason why that is preferred? I thought it would be best to restrict the scope to where the constant is used.

Reply:
given how there is no file scope in Go, I'd tend to agree with Elliot.  Goa in particular requires a huge collection of files in one package.

Reply:
ok, fair enough on that comment. I just find it weird to have a constant here, but if it works for you, then that's fine.